Save messages as drafts when window loses focus

amcdowell226 opened the pull request:
This pull request makes it so any message being written saves as a draft when the user switches tabs.
 
The update_draft function from drafts.ts is called in the on-blur handler for window to prevent message loss. 

Fixes [#35262: Save open draft on window blur](https://github.com/zulip/zulip/issues/35262)

**Screen capture**


https://github.com/user-attachments/assets/f8c50ce6-55e6-4f89-b18f-1979a8653de5

An example of typing a message, switching tabs, and the message saving as a draft.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@amanagr Are you up for reviewing this one?

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -954,5 +955,8 @@ export function initialize(): void {
             window_focused = false;
+
+            // Save the current message as a draft if it exists.
+            drafts.update_draft();
```
This code is correct, but from an abstractions standpoint, I think we should setup an independent handler in `compose_setup.initialize`, so it fits together properly.

amcdowell226 commented:
Thank you for the feedback! 

Changes:
* Corrected my commit message so it follows the guidelines better and adjusted my comment for the window on blur handler
* Moved the handler for window blur to `compose_setup.js`, leaving `unread_ops.ts` as it was before my changes
* Added some handles to stub window objects to `compose.test.cjs` and `compose_video.test.cjs` since it seemed like they can't be read in the test files

I also checked the behavior for when the window is closed and a draft does get saved properly. The code for it was in `reload.ts` like assumed. 

Ready for review!

timabbott commented:
Merged, after fixing some missing `.`s in the commit message, and a typo. It's well worth running a spell-checker and grammar checker on those, as they get read a lot more than one might think :).

Thanks for the contribution @amcdowell226!
